src/schema/authorization/roles.ts

```diff
@@ -0,0 +1,121 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @copyright Copyright (c) 2020-2021 Serlo Education e.V.
+ * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://github.com/serlo-org/api.serlo.org for the canonical source repository
+ */
+import * as R from 'ramda'
+
+import { AuthorizationPayload, Permission, Scope } from '~/authorization'
+
+export enum Role {
+  Guest = 'guest',
+  Login = 'login',
+  Sysadmin = 'sysadmin',
+
+  Moderator = 'moderator',
+  Reviewer = 'reviewer',
+  Architect = 'architect',
+  StaticPagesBuilder = 'static-pages-builder',
+  Admin = 'admin',
+}
+
+interface RoleDefinition {
+  extends?: Role[]
+  permissions?: Permission[]
+}
+
+const roleDefinitions: Record<Role, RoleDefinition> = {
+  [Role.Guest]: {},
+  [Role.Login]: {
+    permissions: [
+      Permission.Thread_CreateThread,
+      Permission.Thread_CreateComment,
+    ],
+  },
+  [Role.Sysadmin]: {
+    extends: [Role.Guest, Role.Login, Role.Moderator],
+  },
+
+  [Role.Moderator]: {
+    permissions: [
+      Permission.Thread_SetThreadArchived,
+      Permission.Thread_SetThreadState,
+      Permission.Thread_SetCommentState,
+    ],
+  },
+  [Role.Reviewer]: {},
+  [Role.Architect]: {},
+  [Role.StaticPagesBuilder]: {},
+  [Role.Admin]: {},
+}
+
+export type RolesPayload = { [scope in Scope]?: Role[] }
+
+export function resolveRolesPayload(
+  payload: RolesPayload
+): AuthorizationPayload {
+  const permissions: AuthorizationPayload = {}
+
+  // Root level
+  const globalRoles = getRolesWithInheritance(payload[Scope.Serlo])
+  let globalPermissions: Permission[] = []
+
+  for (const globalRole of globalRoles) {
+    globalPermissions = R.union(
+      globalPermissions,
+      getPermissionsForRole(globalRole)
+    )
+  }
+  permissions[Scope.Serlo] = globalPermissions
+
+  // Instance level
+  for (const scope of R.values(Scope)) {
+    if (scope === Scope.Serlo) continue
+    let instancedPermissions = [...globalPermissions]
+    const roles = getRolesWithInheritance(payload[scope])
+    for (const role of roles) {
+      instancedPermissions = R.union(
+        instancedPermissions,
+        getPermissionsForRole(role)
+      )
+    }
+    permissions[scope] = instancedPermissions
+  }
+
+  return R.pickBy(R.complement(R.isEmpty), permissions)
+
+  function getPermissionsForRole(role: Role): Permission[] {
+    return roleDefinitions[role].permissions ?? []
```

Comment:
Why not the following? Due to the fact that roles are scoped?

```suggestion
    return getRolesWithInheritance([role]).flatMap(role => role.permissions ?? [])
```

Reply:
Role inheritance is already resolved in the caller function (i.e. `resolveRolesPayload`). This is already slightly optimized since the inheritance is only done once (which makes a difference when multiple of your roles inherit the same roles). I.e.:

- Current implementation:
  - Actual roles: [moderator, sysadmin]
  - Roles with inheritance: [guest, login, moderator, sysadmin]
  - Permissions: permissions(guest) union permissions(login) union permissions(moderator) union permissions(sysadmin)

- Permissions with role inheritance:
  - Actual roles: [moderator, sysadmin]
  - Permissions: permissionsWithInheritance(moderator) union permissionsWithInheritance(sysadmin)
  - Resolved permissions: (permissions(moderator)) union (permissions(guest) union permissions(login) union permissions(moderator) union permissions(sysadmin))

But yeah, probably doesn't really matter either way.